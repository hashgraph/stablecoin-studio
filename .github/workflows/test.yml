name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test