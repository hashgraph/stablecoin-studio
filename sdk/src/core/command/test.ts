/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
