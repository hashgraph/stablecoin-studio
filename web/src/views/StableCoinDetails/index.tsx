import { Box, Flex } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import BaseContainer from '../../components/BaseContainer';
import DetailsReview from '../../components/DetailsReview';
import { SELECTED_WALLET_COIN } from '../../store/slices/walletSlice';

const StableCoinDetails = () => {
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const { t } = useTranslation('stableCoinDetails');

	const getKeyText = (
		key: { key: string; type: string; id: never } | { id: string; key: never; type: never },
	) => {
		if (!key) {
			return t('none');
		}
		if (key.id) {
			return t('smartContract');
		}
		if (key.key === 'same as user') {
			// TODO: check current public key
			return t('currentUser');
		} else {
			return t('otherKey', { key: key.key });
		}
	};

	return (
		<BaseContainer title={t('title')}>
			<Flex justify='center' p={{ base: 4, md: '128px' }} pt={{ base: 4, lg: 14 }}>
				{selectedStableCoin && (
					<Box flex={1} maxW='563px'>
						<DetailsReview
							title={t('subtitle')}
							titleProps={{ fontWeight: 'bold' }}
							contentProps={{ justifyContent: 'start', gap: 4 }}
							details={[
								{
									label: t('tokenId'),
									value: selectedStableCoin?.tokenId,
								},
								{
									label: t('name'),
									value: selectedStableCoin?.name,
								},
								{
									label: t('symbol'),
									value: selectedStableCoin?.symbol,
								},
								{
									label: t('decimals'),
									value: selectedStableCoin?.decimals,
								},
								{
									label: t('totalSupply'),
									value: 0,
								},
								{
									label: t('maxSupply'),
									value: 0,
								},
								{
									label: t('treasuryId'),
									value: selectedStableCoin?.treasuryId,
								},
								{
									label: t('memo'),
									value: selectedStableCoin?.memo,
								},
								{
									label: t('adminKey'),
									value: getKeyText(selectedStableCoin?.adminKey as any), // TODO: remove any when received correct sdk type
								},
								{
									label: t('kycKey'),
									value: getKeyText(selectedStableCoin?.kycKey as any),
								},
								{
									label: t('freezeKey'),
									value: getKeyText(selectedStableCoin?.freezeKey as any),
								},
								{
									label: t('wipeKey'),
									value: getKeyText(selectedStableCoin?.wipeKey as any),
								},
								{
									label: t('supplyKey'),
									value: getKeyText(selectedStableCoin?.supplyKey as any),
								},
							]}
						/>
					</Box>
				)}
			</Flex>
		</BaseContainer>
	);
};

export default StableCoinDetails;
