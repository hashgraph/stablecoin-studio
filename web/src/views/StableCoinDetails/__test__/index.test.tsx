import StableCoinDetails from '../';
import { render } from '../../../test/index';
import translations from '../../../translations/en/stableCoinDetails.json';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();

describe(`<${StableCoinDetails.name} />`, () => {
	beforeEach(() => {});

	test('should render correctly', () => {
		const component = render(<StableCoinDetails />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<StableCoinDetails />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should has subtitle', () => {
		const selectedStableCoin = {
			initialSupply: 0,
			tokenId: '0.0.48162226',
			totalSupply: 0,
			name: 'MIDAS',
			symbol: 'MD',
			decimals: 3,
			id: '0.0.48132286',
			maxSupply: '100000',
			treasuryId: '0.0.48160285',
			memo: 'Hedera Accelerator Stable Coin',
			adminKey: {
				key: 'key',
				type: 'ED25519',
			},
			freezeKey: {
				key: 'key',
				type: 'ED25519',
			},
			wipeKey: {
				id: '0.0.48160285',
			},
			supplyKey: {
				id: '0.0.48160285',
			},
		};

		const store = mockStore({
			wallet: {
				selectedStableCoin,
			},
		});

		const component = render(<StableCoinDetails />, store);
		const subtitle = component.getByTestId('details-review-title');

		expect(subtitle).toHaveTextContent(translations.subtitle);
	});
});
