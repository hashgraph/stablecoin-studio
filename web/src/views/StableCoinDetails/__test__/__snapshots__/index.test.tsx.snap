// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StableCoinDetails /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-llrpa2"
      data-testid="base-container-heading"
    >
      Stable coin details
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    >
      <div
        class="css-t23vsm"
      >
        <div
          class="css-ezt3jl"
        >
          <div
            class="css-13brihr"
          >
            <p
              class="chakra-text css-hduy72"
              data-testid="details-review-title"
            >
              Stable coin basic information
            </p>
            <div
              class="css-0"
            >
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-0"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Token ID
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      0.0.48471242
                    </p>
                    <div
                      class="css-1m8iww1"
                    >
                      <svg
                        class="chakra-icon css-13otjrl"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <rect
                          fill="none"
                          height="256"
                          width="256"
                        />
                        <polyline
                          fill="none"
                          points="168 168 216 168 216 40 88 40 88 88"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                        />
                        <rect
                          fill="none"
                          height="128"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                          width="128"
                          x="40"
                          y="88"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-1"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Name
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      ManuCoin2
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-2"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Symbol
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      MCC2
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-3"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Decimals
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      3
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-4"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Initial supply
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      0
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-5"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Total supply
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      100
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-6"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Max supply
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      0
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-7"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Treasury ID
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      0.0.48450590
                    </p>
                    <div
                      class="css-1m8iww1"
                    >
                      <svg
                        class="chakra-icon css-13otjrl"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <rect
                          fill="none"
                          height="256"
                          width="256"
                        />
                        <polyline
                          fill="none"
                          points="168 168 216 168 216 40 88 40 88 88"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                        />
                        <rect
                          fill="none"
                          height="128"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                          width="128"
                          x="40"
                          y="88"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-8"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Memo
                  </p>
                  <div
                    class="chakra-stack css-n8fv4a"
                  >
                    <div
                      class="chakra-stack css-84zodg"
                    >
                      <p
                        class="chakra-text css-mgupmj"
                      >
                        Proxy contract : 
                      </p>
                      <div
                        class="css-1m8iww1"
                      >
                        <svg
                          class="chakra-icon css-13otjrl"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <rect
                            fill="none"
                            height="256"
                            width="256"
                          />
                          <polyline
                            fill="none"
                            points="168 168 216 168 216 40 88 40 88 88"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="16"
                          />
                          <rect
                            fill="none"
                            height="128"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="16"
                            width="128"
                            x="40"
                            y="88"
                          />
                        </svg>
                      </div>
                    </div>
                    <div
                      class="chakra-stack css-84zodg"
                    >
                      <p
                        class="chakra-text css-mgupmj"
                      >
                        HTS account : 
                      </p>
                      <div
                        class="css-1m8iww1"
                      >
                        <svg
                          class="chakra-icon css-13otjrl"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <rect
                            fill="none"
                            height="256"
                            width="256"
                          />
                          <polyline
                            fill="none"
                            points="168 168 216 168 216 40 88 40 88 88"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="16"
                          />
                          <rect
                            fill="none"
                            height="128"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="16"
                            width="128"
                            x="40"
                            y="88"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-9"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Admin key
                  </p>
                  <div
                    class="css-1p3aqhz"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      Other key: cd7c24...31a622
                    </p>
                    <div
                      class="css-1m8iww1"
                    >
                      <svg
                        class="chakra-icon css-13otjrl"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <rect
                          fill="none"
                          height="256"
                          width="256"
                        />
                        <polyline
                          fill="none"
                          points="168 168 216 168 216 40 88 40 88 88"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                        />
                        <rect
                          fill="none"
                          height="128"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                          width="128"
                          x="40"
                          y="88"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-10"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Kyc key
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      None
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-11"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Freeze key
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      Smart contract
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-12"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Wipe key
                  </p>
                  <div
                    class="chakra-stack css-1c8ysgd"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      Smart contract
                    </p>
                  </div>
                </div>
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-zaaihy"
                  title="divider"
                />
              </div>
              <div
                class="css-0"
              >
                <div
                  class="css-3fqpsw"
                  data-testid="details-review-detail-13"
                >
                  <p
                    class="chakra-text css-bpirlf"
                  >
                    Supply key
                  </p>
                  <div
                    class="css-1p3aqhz"
                  >
                    <p
                      class="chakra-text css-mgupmj"
                    >
                      Other key: cd7c24...31a622
                    </p>
                    <div
                      class="css-1m8iww1"
                    >
                      <svg
                        class="chakra-icon css-13otjrl"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <rect
                          fill="none"
                          height="256"
                          width="256"
                        />
                        <polyline
                          fill="none"
                          points="168 168 216 168 216 40 88 40 88 88"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                        />
                        <rect
                          fill="none"
                          height="128"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="16"
                          width="128"
                          x="40"
                          y="88"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
